File: src/shared/eval/families/king.ts
import { kingAttacks } from "chessops/attacks";
import { SquareSet } from "chessops/squareSet";
import type { Color, Role } from "chessops/types";
import { opposite } from "chessops/util";

import { featureId } from "../features";
import type { FeatureVector } from "../vector";
import type { EvalContext } from "./context";

const KING_ATTACKERS = featureId("kingAttackers");

export const SLOTS = [KING_ATTACKERS];

// What a piece is worth as an attacker near the king — not what it is worth on the board. A queen
// arriving next to the king is the whole attack; a pawn is a nuisance.
const ATTACK_VALUE: Partial<Record<Role, number>> = {
	pawn: 1,
	knight: 2,
	bishop: 2,
	rook: 3,
	queen: 5,
};

function ringOf({ king }: { king: number }): SquareSet {
	return kingAttacks(king).with(king);
}

function attackersOn({ context, color }: { context: EvalContext; color: Color }): number {
	const king = context.position.board.kingOf(color);
	if (king === undefined) return 0;

	const ring = ringOf({ king });
	const enemy = opposite(color);

	let attackers = 0;

	for (const { piece, reach } of context.reach) {
		if (piece.role === "king" || piece.color !== enemy || !reach.intersects(ring)) continue;

		attackers += ATTACK_VALUE[piece.role] ?? 0;
	}

	return attackers;
}

// Pressure around the king, ours minus theirs. A `kingPawnDistance` term sat beside this one and
// was cut: no animal ever weighed it, the lab rated it +30 against bare material — inside the
// ±35 noise band — and it cost a walk of every pawn on the board per node to say it.
export function extractKing({
	context,
	features,
}: {
	context: EvalContext;
	features: FeatureVector;
}): void {
	features[KING_ATTACKERS] =
		attackersOn({ context, color: context.us }) - attackersOn({ context, color: context.them });
}

File: engine/src/eval/king.c
#include "attacks.h"
#include "bitboard.h"
#include "eval.h"
#include "families.h"
#include "feature_ids.h"
#include "position.h"

// What a man is worth as an attacker near the king, not on the board: a queen arriving is the
// whole attack, a pawn a nuisance. Kings are not counted.
static const int ATTACK_VALUE[KING] = {1, 2, 2, 3, 5};

// The weight of every enemy man whose reach touches the king or a square beside him.
static int attackers_on(EvalContext *ctx, Color color) {
	const Position *pos = ctx->pos;
	Square king = bb_first(pos->roles[KING] & pos->colors[color]);
	Bitboard ring = king_attacks(king) | square_bb(king);
	Bitboard enemies = pos->colors[opposite(color)] & ~pos->roles[KING];
	int attackers = 0;
	while (enemies != 0) {
		Square square = bb_pop(&enemies);
		attackers +=
		    (ctx->reach[square] & ring) != 0 ? ATTACK_VALUE[piece_role(pos->board[square])] : 0;
	}
	return attackers;
}

// `extractKing`: danger around our king minus around theirs.
void extract_king(EvalContext *ctx, float *features) {
	eval_walk(ctx);
	features[FEATURE_KING_DANGER] =
	    (float)(attackers_on(ctx, ctx->us) - attackers_on(ctx, ctx->them));
}
